src/ray/core_worker/store_provider/memory_store/memory_store.cc

```diff
@@ -275,7 +277,34 @@ Status CoreWorkerMemoryStore::Get(const std::vector<ObjectID> &object_ids,
   if (should_notify_raylet) {
     RAY_CHECK_OK(raylet_client_->NotifyDirectCallTaskBlocked());
   }
-  bool done = get_request->Wait(timeout_ms);
+
+  bool done = false;
+  bool timed_out = false;
+  Status signal_status = Status::OK();
+  int64_t remaining_timeout = timeout_ms;
+  int64_t iteration_timeout =
+      std::min(timeout_ms, RayConfig::instance().get_timeout_milliseconds());
+  if (timeout_ms == -1) {
+    iteration_timeout = RayConfig::instance().get_timeout_milliseconds();
+  }
+
+  // Repeatedly call Wait() on a shorter timeout so we can check for signals between
+  // calls. If timeout_ms == -1, this should run forever until all objects are
+  // ready or a signal is received. Else it should run repeatedly until that timeout
+  // is reached.
+  while (!(done = get_request->Wait(iteration_timeout)) && !timed_out &&
+         signal_status.ok()) {
+    if (check_signals_) {
+      signal_status = check_signals_();
+    }
+
+    if (remaining_timeout >= 0) {
+      iteration_timeout = std::min(remaining_timeout, iteration_timeout);
+      remaining_timeout -= iteration_timeout;
+      timed_out = remaining_timeout <= 0;
+    }
+  }
+
```

Comment:
this function is getting a bit too long, could we move the new code into a separate function?